File: src/Daemon.cs
//+----------------------------------------------------------------------------
//
//  Copyright (c) 2008 Matthew Loar.
//  See COPYING for license terms.
//
//  Abstract:
//      Implements the daemon which plays tracks.
//
//-----------------------------------------------------------------------------

using System;
using System.Configuration;
using System.Diagnostics;

namespace Neztu
{
  public class Daemon
  {
    public static int Main(string[] args)
    {
      INeztuDatabase database = DatabaseHelper.GetDatabase();
      IRandomSelector randSel = new FullyRandomSelector(database);
      IScheduler s = new FIFOScheduler(database);

      if (args.Length > 0)
      {
        if(args[0] == "-u")
        {
          Vote v = s.GetNext();
          if (v != null)
          {
            database.RemoveVote(v.UserName, v.ReqTrack.TrackId);
            database.AddPending(v.UserName, v.ReqTrack.TrackId);
            Console.WriteLine(v.ReqTrack.Filename);
          }
        }
        else if (args[0] == "-r")
        {
          Track t = randSel.GetRandom();
          Console.WriteLine(t.Filename);
        }
        else if (args[0] == "-n")
        {
          string filename = string.Join(" ", args, 1, args.Length - 1);
          Vote[] votes = database.GetPending();
          foreach (Vote v in votes)
          {
            if (v.ReqTrack.Filename == filename)
            {
              database.RemovePending(v.UserName, v.ReqTrack.TrackId);
              database.AddHistory(v.UserName, v.ReqTrack.TrackId);
              return 0;
            }
          }

          Track t = database.GetTrack(filename);
          database.AddHistory(string.Empty, t.TrackId);
        }
      }
      else
      {
        string mediaPlayerCommand = ConfigurationManager.AppSettings["MediaPlayerCommand"];
        if (string.IsNullOrEmpty(mediaPlayerCommand))
        {
          Console.Error.WriteLine("MediaPlayerCommand not defined in .config file.  Aborting.");
          return 1;
        }

        while (true)
        {
          Track t;

          Vote v = s.GetNext();
          if (v != null)
          {
            t = v.ReqTrack;
            database.RemoveVote(v.UserName, t.TrackId);
            database.AddHistory(v.UserName, t.TrackId);
          }
          else
          {
            t = randSel.GetRandom();
            database.AddHistory(string.Empty, t.TrackId);
          }

          Process.Start(mediaPlayerCommand, "\"" + t.Filename + "\"").WaitForExit();
        }
      }

      return 0;
    }
  }
}

File: Daemon.cpp
//+----------------------------------------------------------------------------
//
//  Copyright (c) 2008 Matthew Loar.
//  See LICENSE for license terms.
//
//  Abstract:
//      Implements the player daemon.
//
//-----------------------------------------------------------------------------

#include <list>
#include <iostream>
#include <stdlib.h>
#include <unistd.h>
#include <signal.h>
#include <string.h>

#include "Database.h"
#include "Scheduler.h"

#include <AL/alut.h>
#include <iostream>

#include "streamer.h"
#include "ogg.h"
#include "mp3.h"

using namespace neztu;

int skip = 0;

void playit(const char *file)
{
    streamer* stream;
    alutInit(NULL, NULL);

    if (mp3_stream::try_open(file))
    {
        stream = new mp3_stream;
    }
    else
    {
        stream = new ogg_stream;
    }

    try
    {
        stream->open(file);

        stream->display();

        if(!stream->playback())
            throw string("File refused to play.");

        while(stream->update() && !skip)
        {
            if(!stream->playing())
            {
                if(!stream->playback())
                    throw string("Playback abruptly stopped.");
                else
                    cout << "Playback was interrupted.\n";
            }
        }

        skip = 0;
    }
    catch(string error)
    {
        cout << error;
    }

    delete stream;

    alutExit();
}

void sig_handler(int signal)
{
    if (signal == SIGUSR1)
        skip = 1;
}

int main(int argc, char* argv[])
{
  Configuration config("/etc/neztu.conf");
  Database db(config);
  Scheduler scheduler(config, db);

  // XXX: use getopt or something
  if ((argc < 2 || strcmp(argv[1], "-d")) && daemon(0, 0))
  {
    std::cerr << "Could not fork" << std::endl;
    return 1;
  }

  struct sigaction act;
  memset(&act, 0, sizeof(struct sigaction));
  act.sa_handler = sig_handler;
  if (sigaction(SIGUSR1, &act, NULL))
  {
      std::cerr << "sigaction failed" << std::endl;
      return 1;
  }

  for (;;)
  {
    const Vote &v = scheduler.GetNext();
    if (!v.Timestamp.empty())
    {
      printf("Playing %s\n", v.ReqTrack.Title.c_str());
      db.RemoveVote(v.UserName, v.ReqTrack.TrackId);
      db.AddHistory(v.UserName, v.ReqTrack.TrackId);

      printf("%s\n", v.ReqTrack.Filename.c_str());
      playit(v.ReqTrack.Filename.c_str());
    }
    else
    {
      std::vector<Track> tracks;
      printf("No votes...playing random song\n");
      db.GetRandomTracks(&tracks, 1);
      db.AddHistory("neztu", tracks[0].TrackId);
      printf("%s\n", tracks[0].Filename.c_str());
      playit(tracks[0].Filename.c_str());
    }
  }

  return 0;
}
